external/swloop/swloop.cpp

```diff
@@ -0,0 +1,533 @@
+            if (params.formfact) {
+                expF *= FF(i1, jj); }
+            for (size_t i2=0; i2<3; i2++) {
+                // Don't know why we have to use the reverse of the Matlab code here and in
+                // V.tranpose() below instead of V.adjoint() - but otherwise get wrong intensities...
+                zedExpF(i1, i2) = std::conj(zed(i2, i1)) * expF;
+                zedExpF(i1 + params.nMagExt, i2) = zed(i2, i1) * expF;
+            }
+        }
+        V = V.transpose().eval();
+        Eigen::MatrixXcd VExp(3, nHam);
+        for (size_t i1=0; i1<3; i1++) {
+            VExp(i1, Eigen::all) = V * zedExpF(Eigen::all, i1);
+        }
+        if (params.incomm) {
+            size_t kk = jj % nHklT;
+            if (kk < nHklI) {
+                K = K1;
+            } else if (kk < 2*nHklI) {
+                K = K2;
+            } else {
+                K = cK1;
+            }
+        }
+        for (int i1=0; i1<nHam; i1++) {
+            Eigen::Map<Eigen::Matrix3cd> Sab(Sab_ptr);
+            Sab = VExp(Eigen::all, i1) * VExp(Eigen::all, i1).adjoint();
+            if (params.incomm) {
+                if (params.helical) {
+                    // integrating out the arbitrary initial phase of the helix
+                    Eigen::Matrix3cd tmp = (nx * Sab * nx) - ((K2 - m1) * Sab * K2) - (K2 * Sab * (2*K2 - m1));
+                    Sab = 0.5 * (Sab - tmp);
+                }
+                Sab = Sab * K;
+            }
+            if (params.nTwin > 1) {
+                // Rotates the correlation function by the twin rotation matrix
+                size_t iT = jj / nHklT;
+                Eigen::Map<const Eigen::Matrix3d> rotC(inputs.rotc + iT*9, 3, 3);
+                Sab = rotC * Sab * rotC.transpose().eval();
+            }
+            Sab_ptr += 9;   // This trick only works for column-major data layouts!
+        }
+    }
+}
+
+template <typename T> T getVal(const mxArray *d) { return (T)*(mxGetDoubles(d)); }
+template <> bool getVal(const mxArray *d) { return (bool)*((mxLogical*)mxGetData(d)); }
+template <typename T>
+T getField(const mxArray *data_ptr, size_t idx, const char *fieldname, T default_val) {
+    mxArray *field = mxGetField(data_ptr, idx, fieldname);
+    if (field == nullptr) {
+        return default_val;
+    }
+    return getVal<T>(field);
+}
+
+void checkDims(std::string varname, const mxArray *data, size_t dim1, size_t dim2) {
+    const mwSize *dims = mxGetDimensions(data);
+    if ((size_t)dims[0] != dim1 || (size_t)dims[1] != dim2) {
+        throw std::runtime_error("Input " + varname + " has the incorrect size");
+    }
+}
+
+void mexFunction( int nlhs, mxArray *plhs[], int nrhs, const mxArray *prhs[] )
+{
+    if (nrhs < 15) {
+        throw std::runtime_error("swloop: Requires 15 arguments");
+    }
+    err_flag.store(0, std::memory_order_release);
+    size_t nThreads = getField<size_t>(prhs[0], 0, "nThreads", 8);
```

Comment:
What happens if the user requests more threads than they have? Seems like the default 8 (if that is the default?) is quite high!

Reply:
Ah, the `mex` file always spawns as many threads as requested (e.g. defaults to `8`). If there are less cores than the number of threads then each core will run multiple threads, context-switching between them which will result in a loss of efficiency.

We can set a default which is the same as the number of cores on a user's system... there are two ways to do this - either in the `m`-file code or in the C++ code, but the problem with the C++ code is that it would need to be OS-dependent (and I'm not exactly sure how to do it for Mac OSX).

Reply:
Actually there is a OS-independent way of getting the number of threads in C++: [std::thread::hardware_concurrency](https://en.cppreference.com/w/cpp/thread/thread/hardware_concurrency). I'll use this in the mex code and have the m-file code set `nTheads=-1` by default to mean "automatic".